### PR TITLE
Average rating

### PR DESCRIPTION
**Pull request by dag13**
Create updateAverageRating method and create test cases for it.

### CODE REVIEW

**Review by TylerHernandez — DISMISSED** [review]

File: src/main/java/com/google/sps/data/Drink.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    return null;
+  }
+
+  // TODO: DATASTORE
+
+  public double updateAverageRating(String id, double newRating) {
+    Point2D ratings = storeIDToAvgRating.get(id);
+    double numRatings = ratings.getX();
+    double averageRating = ratings.getY();
+    double errorReturnValue = -Double.MAX_VALUE;
```
Comment (yjzhang111) [inline src/main/java/com/google/sps/data/Drink.java]: Did you consider the case where storeID has not yet been saved in the map?

File: src/main/java/com/google/sps/data/Drink.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    if (numRatings == 0.0) {
+      ratings.setLocation(1.0, newRating);
+      storeIDToAvgRating.put(id, ratings);
+      return newRating;
+    }
+
+    averageRating = ((averageRating * numRatings) + newRating) / (numRatings + 1);
+    averageRating = roundToOneDecimalPlace(averageRating);
```
Comment (yjzhang111) [inline src/main/java/com/google/sps/data/Drink.java]: nit: rounding in every single calculation could lose accuracy. Storing the actual value and rounding when displaying on the page could be another approach.

File: src/test/java/com/google/sps/data/DrinkTest.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+  }
+
+  @Test
+  public void testUpdateAverageRatingG() {
+    double newRating = 0.0;
+    double actual = drink.updateAverageRating(ID_G, newRating);
+    double expected = ERROR_RETURN_VALUE;
+    Assert.assertEquals(expected, actual, epsilon);
+  }
+}
```
Comment (yjzhang111) [inline src/test/java/com/google/sps/data/DrinkTest.java]: A potential test case (also mentioned previously) is to test updating average rating for a new storeID.

**Review by yjzhang111 — COMMENTED** [review]
Commented on an edge case in updateAverageRating() method, PTAL.

File: src/main/java/com/google/sps/data/Drink.java (hunk 2 above)
Comment (dag13) [reply src/main/java/com/google/sps/data/Drink.java]: Ok I'll make that change.

File: src/test/java/com/google/sps/data/DrinkTest.java (hunk 3 above)
Comment (dag13) [reply src/test/java/com/google/sps/data/DrinkTest.java]: Ok I'll make a test case like that

**Comment by albert-g3** [comment]
@dag13 update this PR with latest code from main branch

File: src/main/java/com/google/sps/data/Drink.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+  // TODO: DATASTORE
+
+  public double updateAverageRating(String id, double newRating) {
+    Point2D ratings = storeIDToAvgRating.get(id);
+    double numRatings = ratings.getX();
+    double averageRating = ratings.getY();
+    double errorReturnValue = -Double.MAX_VALUE;
+
+    if (newRating == 0.0) {
```
Comment (albert-g3) [inline src/main/java/com/google/sps/data/Drink.java]: How can this happen?

File: src/main/java/com/google/sps/data/Drink.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+package com.google.sps.data;
+
+import java.awt.geom.Point2D;
+import java.util.HashMap;
+
+public class Drink {
+
+  private final String name;
+  private final String id;
+  public HashMap<String, Point2D> storeIDToAvgRating;
```
Comment (albert-g3) [inline src/main/java/com/google/sps/data/Drink.java]: using Point2D is hacky here.  What are the reasons for it? Is there something else you can use?

File: src/main/java/com/google/sps/data/Drink.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  public String getName() {
+    return this.name;
+  }
+
+  public String getID() {
+    return this.id;
+  }
+
+  public double roundToOneDecimalPlace(double num) {
+    if (num < 1.0) return 0.0;
```
Comment (albert-g3) [inline src/main/java/com/google/sps/data/Drink.java]: so for example 0.95 would be rounded to 0.0? The name of the method implies it to it should be 1.0

**Review by albert-g3 — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/data/Drink.java (hunk 4 above)
Comment (dag13) [reply src/main/java/com/google/sps/data/Drink.java]: It can't I just put it there just in case.

File: src/main/java/com/google/sps/data/Drink.java (hunk 5 above)
Comment (dag13) [reply src/main/java/com/google/sps/data/Drink.java]: I can use something else.

**Comment by mutisog** [comment]
Why is this pull request closed?

**Comment by dag13** [comment]
This pull request is closed because the changes made to the Drink class wouldn't compile or make sense without a DrinkDAO. So the DrinkDAO pull request accomplishes everything the two closes pull requests aimed to do.
